### PR TITLE
feat: Genesis Types

### PR DESCRIPTION
**Pull request by refcell**
### Description

Introduces an `op-alloy-genesis` crate similar to [`alloy-genesis`](https://docs.rs/alloy-genesis/latest/alloy_genesis/) that contains Optimism genesis types.

This is part of an effort to remove _both_ `superchain-primitives` and `kona-primitives` to place all optimism core types in `op-alloy` crates.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -27,15 +27,20 @@ option_if_let_else = "warn"
 all-features = true
 rustdoc-args = ["--cfg", "docsrs"]
 
+[patch.crates-io]
+alloy-eips = { git = "https://github.com/alloy-rs/alloy", branch = "main" }
+
```
Comment (refcell) [inline Cargo.toml:32]: TODO: remove this once [`BlockNumHash`](https://docs.rs/alloy-eips/latest/alloy_eips/eip1898/type.BlockNumHash.html) has serde support.

File: crates/genesis/src/params.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+/// Returns the Canyon [BaseFeeParams] for the given chain id.
+pub const fn canyon_base_fee_params(chain_id: u64) -> BaseFeeParams {
+    match chain_id {
+        10 => OP_CANYON_BASE_FEE_PARAMS,
+        11155420 => OP_SEPOLIA_CANYON_BASE_FEE_PARAMS,
+        8453 => OP_CANYON_BASE_FEE_PARAMS,
+        84532 => BASE_SEPOLIA_CANYON_BASE_FEE_PARAMS,
+        _ => OP_CANYON_BASE_FEE_PARAMS,
+    }
+}
```
Comment (mattsse) [inline crates/genesis/src/params.rs:25]: I'd like to move this below the constants

File: crates/genesis/src/rollup.rs [hunk 3]
```diff
@@ -0,0 +1,708 @@
...
+            protocol_versions_address: Address::ZERO,
+            superchain_config_address: None,
+            blobs_enabled_l1_timestamp: None,
+            da_challenge_address: None,
+        }
+    }
+}
+
+/// Loads the rollup config for the OP-Stack chain given the chain config and address list.
+pub fn load_op_stack_rollup_config(chain_config: &ChainConfig) -> RollupConfig {
```
Comment (mattsse) [inline crates/genesis/src/rollup.rs:177]: can we make this a function of chainconfig instead?

**Review by mattsse — APPROVED** [review]
didn't review the function impls too closely, but overall lgtm, pending some suggestions

File: crates/genesis/src/rollup.rs (hunk 3 above)
Comment (refcell) [reply crates/genesis/src/rollup.rs:177]: Definitely good idea.

File: crates/genesis/src/chain.rs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+pub struct AltDAConfig {
+    /// AltDA challenge address
+    pub da_challenge_address: Option<Address>,
+    /// AltDA challenge window time (in seconds)
+    pub da_challenge_window: Option<u64>,
+    /// AltDA resolution window time (in seconds)
+    pub da_resolve_window: Option<u64>,
+}
+
+/// Hardfork configuration.
```
Comment (emhane) [inline crates/genesis/src/chain.rs:33]: this would be nice to have for alloy too, could you pls open an issue

**Review by emhane — APPROVED** [review]
nice!

**Pull request merged**